---
title: "SLE712 Assessment 3- Part 2-Determine the limits of BLAST"
author: "Yasodha Jayawardena & Dhanusha Weerawardana "
output: html_document
theme: cosmo 
---

# Introduction
BLAST is a program or algorithm which is important combinaion of biology and technology to compare biological sequences including, amino acid sequences, and DNA and RNA nucleotied sequences. During this assesment, students able to gain knowladge on BLAST program as wel as able to determine the limits of BLAST. Therefore, in order to compare the sequences, E.Coli gene sequence and Student56 gene sequence have used through out the process. 


# Prerequisite libraries
There are three packages that we will need to get things working. 
`seqinr` package designed for sequence data procesing and analysis
`R.utils` package designed for extract data from compressed files
`rBLAST` package provide R interface for BLAST search
Additionally, we source two functions which supplied for this practical to perform an analysis into the limits of BLAST.

```{r,chunk1,echo=TRUE}
#libraries that we need
library("seqinr")
library("R.utils")
library("rBLAST")
```

# Question 1-Download the whole set of E. coli gene DNA sequences and use gunzip to decompress to create a blast database.

Here we are used the `download.file()` command to obtain the E.coli dataset from ensemble FTP page and then we are decompress the file using gunzip command. `makeblast()` function use to creat a BLAST database.
*Number of sequences are present in the E.coli set= 4140

```{r,chunk2,echo=TRUE}
#Download the E.coli gene sequency from the Ensemble FTP page
download.file("ftp://ftp.ensemblgenomes.org/pub/bacteria/release-42/fasta/bacteria_0_collection/escherichia_coli_str_k_12_substr_mg1655/cds/Escherichia_coli_str_k_12_substr_mg1655.ASM584v2.cds.all.fa.gz", destfile = "Escherichia_coli_str_k_12_substr_mg1655.ASM584v2.cds.all.fa.gz")
#use gunzip to uncompress the file
gunzip("Escherichia_coli_str_k_12_substr_mg1655.ASM584v2.cds.all.fa.gz",overwrite=TRUE)
#create a blast database with makeblast function
makeblastdb("Escherichia_coli_str_k_12_substr_mg1655.ASM584v2.cds.all.fa",dbtype = "nucl","-parse_seqids")
```
# Question 2-Download the sample fasta sequences and read the sequence.

Here we are used the `download.file()` command to obtain the sample file and then read it into R. The `str` use to determine data structure. Next select allocated sequence 56 and read it as "myEcoli" and check its structure. 
Determine length and propotion of GC bases in allocated sequence by using `getLength()` command and `GC()` command respectively.
*Length=276 bp
*The propotion of GC bases= 0.5326087

```{r,chunk3,echo=TRUE}
#Download sample file
download.file("https://raw.githubusercontent.com/markziemann/SLE712_files/master/bioinfo_asst3_part2_files/sample.fa", destfile = "sample.fa")
#Read file into R as Ecoli
Ecoli<-read.fasta("sample.fa")
str(Ecoli) #check data structure
#Read my allocated sequence (56) into R as myEcoli
myEcoli<-Ecoli[[56]]
#Lets check if the data has been imported properly
myEcoli
str(myEcoli)
#calculate length for my allocated gene
seqinr::getLength(myEcoli)
#calculate GC content for my allocated gene
seqinr::GC(myEcoli)
```
# Question 3-Use Blast to identify what E. coli gene your sequence matches best and show a table of percent identity, E-value and bit scores for top3 hits.

In order to creat a BLAST database, run the BLAST by the `download.file(location of the file)` command. Then `res<-` function can be applied to test the command and `head()` and `str` commands can be used to arrenge the table acoording to structure. 

```{r,chunk4,echo=TRUE}
#sourced the function that ran blast searches in R
download.file("https://raw.githubusercontent.com/markziemann/SLE712_files/master/bioinfo_asst3_part2_files/mutblast_functions.R",destfile = "mutblast.R")
source("mutblast.R")
#test the function
res<-myblastn_tab(myseq = myEcoli,db= "Escherichia_coli_str_k_12_substr_mg1655.ASM584v2.cds.all.fa")
#have a look at the blast results
head(res)
str(res)
```

The first top 3 hits can be obtain by creating a character `hits` and applying running command for the hits.

```{r,chunk5,echo=TRUE}
###first three hits
hits <-as.character(res$sseqid[1:3])
hits 
```

To identify the best matches between two gene sequences `BLAST` fuction can be used. Therefore, `db<-read.fast` command can be applied. The `db` is function for the database.

```{r,chunk6,echo=TRUE}
#identify what E.coil gene matches best
db<-read.fasta("Escherichia_coli_str_k_12_substr_mg1655.ASM584v2.cds.all.fa")
str(db)
head(names(db))
head(db)
```

# Question 4-Make a set number of point mutations to your sequence of interest and check the number of mismatchesbetween the original and mutated sequence.

In this question, first `seqinr` package has to be run and it is a Exploratory data analysis and data visualization for biological sequence and then created a blast database by the command `makeblast`.
Thereafter the `cat` command can be applied to produce the out by converting argument in to a character vectors.
Then we have given the command `mutator` to extract mutation with 30 substitutions.
```{r,chunck7,echo=TRUE}
seqinr::write.fasta(myEcoli,names="myEcoli",file.out = "myEcoli.fa")
makeblastdb("myEcoli.fa", dbtype = "nucl","-parse_seqids")
res <- myblastn(myseq = myEcoli, db= "myEcoli.fa")
res
cat(res,fill=TRUE)
###create a mutated copy with 30 substitutions
myEcolimutator<- mutator(myEcoli,30)
res <- myblastn_tab(myseq = myEcolimutator, db= "myEcoli.fa")
res
```

# Question 5-Determine the number and proportion of sites that need to be altered to prevent the BLAST search from matching the gene of origin.

```{r,chunck8,echo=TRUE}
###test with mismatches
myEcolimutator <- mutator(myEcoli,20)
res<-myblastn_tab(myseq = myEcolimutator,db= "myEcoli.fa")
res
myEcolimutator <- mutator(myEcoli,30)
res<-myblastn_tab(myseq = myEcolimutator,db= "myEcoli.fa")
res
myEcolimutator <- mutator(myEcoli,40)
res<-myblastn_tab(myseq = myEcolimutator,db= "myEcoli.fa")
res
myEcolimutator <- mutator(myEcoli,50)
res<-myblastn_tab(myseq = myEcolimutator,db= "myEcoli.fa")
res
myEcolimutator <- mutator(myEcoli,60)
res<-myblastn_tab(myseq = myEcolimutator,db= "myEcoli.fa")
res
myEcolimutator <- mutator(myEcoli,65)
res<-myblastn_tab(myseq = myEcolimutator,db= "myEcoli.fa")
res
```

